Feat: stdlib: adds `system.string.setLenUninit`

ZoomRmc opened the pull request:
Adds `system.setLenUninit` for the `string` type. Allows setting length without initializing new memory on growth.

- Required for a follow-up to #15951
- Accompanies #22767 (ref #19727) but for strings
- Expands `stdlib/tstring` with tests for `setLen` and `setLenUninit`

ZoomRmc commented:
`systr.setLengthStrUninit` fails with `mm:refc` on Linux.
Seems my naïve approach is wrong.
Can this be made to work without introducing new compiler procs?

arnetheduck commented:
~~the same for seq would be nice, ie strings are pretty niche at the end of the day.~~
never mind, forgot that it was already fixed :)

arnetheduck commented:
ping, would be nice to see this PR merged

ZoomRmc commented:
I think I was stuck on not being sure if it's done and both string versions actually work, since some of CI errors at the time were unrelated.

Those string/seq impls is a mess of conditionals and I don't trust locally-run tests.

ZoomRmc commented:
https://github.com/nim-lang/Nim/blob/2e45f6149f0f2feafaf750d3c693d94cb43a5c48/lib/system.nim#L2316-L2326

Adding a `hasAlloc` check and and `else-`downgrade to `setLen` for the old string branch was necessary for this to compile. I'm not sure if it's correct as I'm not sure what constitutes the `string` type at this point.

Araq commented on lib/system/sysstr.nim:
```diff
@@ -156,7 +156,6 @@ proc addChar(s: NimString, c: char): NimString =
       result = rawNewStringNoInit(r)
       result.len = s.len
       copyMem(addr result.data[0], unsafeAddr(s.data[0]), s.len+1)
-      result.reserved = r
```
Why was this removed?

ZoomRmc replied:
It's set at line 156 by the `rawNewStringNoInit` call ([line 53](https://github.com/nim-lang/Nim/blob/2e45f6149f0f2feafaf750d3c693d94cb43a5c48/lib/system/sysstr.nim#L53)). `copyMem` call at line [156](https://github.com/nim-lang/Nim/blob/2e45f6149f0f2feafaf750d3c693d94cb43a5c48/lib/system/sysstr.nim#L156) touches data only, so I don't see the reason to set `reserved` again.

Araq replied:
Ok.

Araq commented on lib/system/sysstr.nim:
```diff
@@ -202,7 +201,6 @@ proc resizeString(dest: NimString, addlen: int): NimString {.compilerRtl.} =
     result = rawNewStringNoInit(sp)
     result.len = dest.len
     copyMem(addr result.data[0], unsafeAddr(dest.data[0]), dest.len+1)
-    result.reserved = sp
```
Why was this removed?

ZoomRmc replied:
Same thing, `rawNewStingNoInit` sets the `reserved` field for the returned NimString.

Araq replied:
Ok.

Araq commented on lib/system/sysstr.nim:
```diff
@@ -220,22 +218,42 @@ proc appendChar(dest: NimString, c: char) {.compilerproc, inline.} =
-      return s
-    else:
-      result = mnewString(n)
+    return if n == 0: s else: mnewString(n)
```
This is not what the original code did and a distraction. Bring back the old version.

ZoomRmc replied:
Again, if we don't return early from this case (and checking for nil is cleraly an "early return" kind of branch),
we're setting `len` and the nul-byte for the result again at lines 229-230 https://github.com/nim-lang/Nim/blob/2e45f6149f0f2feafaf750d3c693d94cb43a5c48/lib/system/sysstr.nim#L229-L230.

`mnewString` sets the `len` and the null-byte on each call for the returned string (though, the null-byte implicitly, by `zeroMem` call):
https://github.com/nim-lang/Nim/blob/2e45f6149f0f2feafaf750d3c693d94cb43a5c48/lib/system/sysstr.nim#L62-L65

Original code for the `s == nil and n != 0` went like this:

```nim
# start of the mnewString call
result = rawNewStringNoInit(n)
result.len = n
zeroMem(addr result.data[0], n + 1) # n (or len in the proc) + 1 sets the null-byte
# after-coditional tail of setLengthStr
result.len = n
result.data[n] = '\0'
```
So this is redundant.

How about I separate the `nil` case into separate if statement to signify it's an early-return case?

ZoomRmc replied:
I mean, it could look like this:
```nim
proc setLengthStr(s: NimString, newLen: int): NimString {.compilerRtl.} =
  let n = max(newLen, 0)
  if s == nil: # early return check
    if n == 0:
      return s
    else:
      return mnewString(n) # sets everything required
  if n <= s.space:
    result = s
  else:
    let sp = max(resize(s.space), n)
    result = rawNewStringNoInit(sp) # len and null-byte not set
    copyMem(addr result.data[0], unsafeAddr(s.data[0]), s.len)
    zeroMem(addr result.data[s.len], n - s.len)
  result.len = n
  result.data[n] = '\0'
```

This is more explicit and in terms of logic completely equal to the original version, barring reduntant field touches, but I'm not totally sure what's "distracting" to you.

ZoomRmc replied:
I think my previous version is more succinct, but indeed it might be better to change the `elif` at ln.227 to a separate `if` to separate the early-return statement.

Araq replied:
It's distracting as it's a change that has nothing to do with your feature addition.

Araq commented:
Apart from my code review remarks, seems fine.

ZoomRmc commented:
So I went with expanding that early return conditional at line 232 a bit and added some internal docs. If it's still not up your alley I can change the code to whatever.

If it's ok as is, do you want me to squash it?

BTW, you can't currently build docs with `--docInternal` for `system` due to assertion defect in `semtypes`:
```
...assertions.nim(34)       raiseAssert
Error: unhandled exception: semtypes.nim(2471, 7) `c.graph.sysTypes[tySequence] == nil`  [AssertionDefect]```

ZoomRmc committed:
Feat: stdlib: adds `system.string.setLenUninit`

- Required for a follow-up to #15951
- Accompanies #19727 but for strings
- Expands `stdlib/tstring` with tests for `setLen` and `setLenUninit`

Feat: stdlib: adds `system.string.setLenUninit`

- Required for a follow-up to #15951
- Accompanies #19727 but for strings
- Expands `stdlib/tstring` with tests for `setLen` and `setLenUninit`

ZoomRmc committed:
SSO strings `setLenUninit`

ZoomRmc commented on lib/system/strs_v3.nim:
```diff
@@ -574,11 +586,19 @@ proc setLengthStrV2(s: var SmallString; newLen: int) {.compilerRtl.} =
         zeroMem(addr s.more.data[curLen], newLen - curLen)
       s.more.data[newLen] = '\0'
-      s.more.fullLen = newLen
+
```
Similarly to the previous attempt for this PR, this line removed as redundant, already performed by `ensureUniqueLong` at l590.

ZoomRmc commented:
Tests updated to support SSO. Matrix now **excludes** `--backend:cpp -d:nimsso` due to buggy codegen (#25735) and some of the tests relying on mutable char access are now conditional.

Araq commented on lib/system/strs_v3.nim:
```diff
@@ -496,12 +496,16 @@ proc mnewString(len: int): SmallString {.compilerproc.} =
 
-proc setLengthStrV2(s: var SmallString; newLen: int) {.compilerRtl.} =
-  ## Sets the length of s to newLen, zeroing new bytes on growth.
+proc setLengthStr(s: var SmallString; newLen: int; zeroing: static bool) =
```
```suggestion
proc setLengthStr(s: var SmallString; newLen: int; zeroing: bool) =
```

Code-size is more important than avoiding 2 runtime branches that are 100% predictable anyway.

ZoomRmc replied:
Ok, I'll try to do it this way next time.

My reasoning was, most of the time I've seen this solved via `…Imp` templates which, of course, duplicate code the same way, so I though it's the preferred outcome of the two.

Araq committed:
Apply suggestion from @Araq

make `zeroing` an rt-argument

ZoomRmc commented:
Test fails in `tests/destructor/trecursivedestructor.nim`, looks unrelated - hits neighbouring PRs too.

ZoomRmc commented:
Now it's the timeout again.
